src/shared/switch-root.c

```diff
@@ -92,35 +95,49 @@ int switch_root(const char *new_root,
          * and switch_root() nevertheless. */
         (void) base_filesystem_create(new_root, UID_INVALID, GID_INVALID);
 
-        if (chdir(new_root) < 0)
+        if (fchdir(new_root_fd) < 0)
                 return log_error_errno(errno, "Failed to change directory to %s: %m", new_root);
 
         /* We first try a pivot_root() so that we can umount the old root dir. In many cases (i.e. where rootfs is /),
          * that's not possible however, and hence we simply overmount root */
-        if (pivot_root(new_root, resolved_old_root_after) >= 0) {
-
-                /* Immediately get rid of the old root, if detach_oldroot is set.
-                 * Since we are running off it we need to do this lazily. */
-                if (unmount_old_root) {
-                        r = umount_recursive(old_root_after, MNT_DETACH);
-                        if (r < 0)
-                                log_warning_errno(r, "Failed to unmount old root directory tree, ignoring: %m");
+        if (resolved_old_root_after)
+                r = RET_NERRNO(pivot_root(".", resolved_old_root_after));
+        else {
+                r = RET_NERRNO(pivot_root(".", "."));
+                if (r >= 0) {
+                        /* This worked? Yay! Now switch back to old root fs as current dir, so that we
+                         * unmount the upper of the two stacked file systems */
+                        if (fchdir(old_root_fd) < 0)
+                                return log_error_errno(errno, "Failed to change directory back to old root: %m");
+
+                        /* Now unmount the upper of the two stacked file systems */
+                        if (umount2(".", MNT_DETACH) < 0)
+                                return log_error_errno(errno, "Failed to unmount the old root: %m");
+
+                        if (fchdir(new_root_fd) < 0)
```

Comment:
Why these two fchdir()s are required? That's not mentioned in pivot_root(2).

Reply:
we can't unmount a dir that is still the cwd of some process...

will extend the comment.

Reply:
So yeah, this turns out not to be necessary after all.

To my defense, I cargo culted this from out other switch root code: https://github.com/systemd/systemd/blob/main/src/shared/mount-util.c#L434

(we should unify the two implementations, but not today).

will update this patch too